apps/rush-lib/src/logic/ProjectTaskSelector.ts

```diff
@@ -1,220 +1,170 @@
+      return [phase, project];
+    }
 
-        if (phase.dependencies?.upstream) {
-          for (const dependencyPhaseName of phase.dependencies.upstream) {
-            const dependencyPhase: IPhase = this._getPhaseByName(dependencyPhaseName);
-            for (const dependencyProject of getSelectedDependencies(project)) {
-              const dependencyTaskName: string = this._getPhaseDisplayNameForProject(
-                dependencyPhase,
-                dependencyProject
-              );
+    /**
+     * Enumerates the declared dependencies of the task in (phase * project) key space
+     * task_ordinal = (project_count * phase_index) + project._index
+     */
+    function* getRawDependencies(taskKey: ITaskKey): Iterable<ITaskKey> {
+      const [
+        {
+          phaseDependencies: { self, upstream }
+        },
+        project
+      ] = getPhaseAndProject(taskKey);
+
+      for (const phase of self) {
+        // Different phase, same project
+        yield getTaskKey(phase, project);
+      }
 
-              dependencyTasks.add(dependencyTaskName);
+      if (upstream.size) {
+        const { dependencyProjects } = project;
+        if (dependencyProjects.size) {
+          for (const phase of upstream) {
+            for (const dependencyProject of dependencyProjects) {
+              yield getTaskKey(phase, dependencyProject);
             }
           }
         }
-
-        taskDependencies.set(taskName, dependencyTasks);
       }
     }
 
-    for (const [taskName, dependencies] of taskDependencies) {
-      taskCollection.addDependencies(taskName, dependencies);
-    }
-
-    return taskCollection;
-  }
+    const filteredDependencyCache: Map<ITaskKey, ITaskDependencies> = new Map();
+    function getFilteredDependencies(node: ITaskKey): ITaskDependencies {
+      const cached: ITaskDependencies | undefined = filteredDependencyCache.get(node);
+      if (cached) {
+        return cached;
+      }
 
-  private _getPhaseByName(phaseName: string): IPhase {
-    const phase: IPhase | undefined = this._phases.get(phaseName);
-    if (!phase) {
-      throw new Error(`Phase ${phaseName} not found`);
-    }
+      const dependencies: ITaskDependencies = {
+        tasks: undefined,
+        isCacheWriteAllowed: taskByKey.has(node)
+      };
 
-    return phase;
-  }
+      filteredDependencyCache.set(node, dependencies);
 
-  private _getPhaseDisplayNameForProject(phase: IPhase, project: RushConfigurationProject): string {
-    if (phase.isSynthetic) {
-      // Because this is a synthetic phase, just use the project name because there aren't any other phases
-      return project.packageName;
-    } else {
-      const phaseNameWithoutPrefix: string = phase.name.substring(RushConstants.phaseNamePrefix.length);
-      return `${project.packageName} (${phaseNameWithoutPrefix})`;
-    }
-  }
+      for (const dep of getRawDependencies(node)) {
```

Comment:
If I change the signature to `const [depKey, depProject, depPhase] of getRawDependencies(project, phase)` and likewise add the `project, phase` parameters to `getFilteredDependencies(key, project, phase)` I can remove the need to ever deserialize the keys, which is probably a performance win.

Reply:
Though since this whole execution takes on the order of 25 ms for 2 phases with 600 projects, might not be terribly relevant.